src/transformers/trainer.py

```diff
@@ -1588,7 +1592,14 @@ def prediction_loop(
             prediction_loss_only if prediction_loss_only is not None else self.args.prediction_loss_only
         )
 
+        if self.args.deepspeed and not self.args.do_train:
+            # In the future we probably can run deepspeed for inference too, but this will require some thinking about how to best run it - since while it works DeepSpeed wasn't designed for inference
+
+            # since we have to postpone model.to() till training for DeepSpeed, if there was no training, we must put the model on the right device
+            self.model = self.model.to(self.args.device)
```

Comment:
Mmm, this means a future training might have it on the device already, now? Maybe we should just put on the device the model used (so `model = self.model.to(self.args.device)` but not stored in `self.model`.

Reply:
this is the case where one bypasses the training stage. Remember last PR here had to make a special case for deepspeed not to preload on device so that it could load a model in fp16?

Next I'm experimenting with DeepSpeed for inference only, so this will change again. But for now this is a bug fix.

Reply:
Yes I understand that. But what if someone does:
```
trainer = Trainer(...)
trainer.evaluate()
trainer.train()
```
(agreed it would be weird but trying to have the bug fix be general)

Reply:
This is yet another combination I haven't thought of. Thank you for thinking of it, @sgugger 

As I mentioned I'm already working on DeepSpeed for inference so this code will change again shortly. And if I manage to do it - this code will be replaced with deepspeed_init and no switching to device at all. So this area is a wip and this PR is a temporary patch.

So do let me know whether you prefer a more general fix or hopefully today/tomorrow I will have a new version if DeepSpeed supports that - I just started working on it and I think in the worst case if it doesn't let me init it for inference (i.e. w/o optimizer/scheduler) I'll just init DeepSpeed as I'd for training if it's not supporting that at the moment, so really it'd be the same as train. Down the road as DeepSpeed avails itself for inference it'll then improve again. That's the plan at the moment.

And yes, I need to test all these different variations you're pointing at.

Reply:
Ok for the quick hotfix then! Just want to make sure the proper fix down the road supports all kinds of combination of train/eval.

Reply:
yes, let's merge it and I will work on the new tests location and then add new tests for all the different combinations.